### PR TITLE
Refactor of Physical vs Analytical Push and Pull

### PR DESCRIPTION
**Pull request by pawelbaran**
~~### PLEASE DO NOT MERGE YET UNDER ANY CONDITIONS!~~

### Issues addressed by this PR

~~**This PR is still in progress!**~~

Issues possibly to be resolved with this PR (to be reviewed when ready to merge):
- [ ] #20
- [ ] #90
- [ ] #131
- [ ] #161
- [ ] #171
- [ ] #174
- [ ] #227
- [ ] #238
- [ ] #274
- [ ] #316
- [ ] #317 

### Test files

@ZiolkowskiJakub @michaldengusiak to test the BEnv side.

General:
- [physical pull/push](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2F%5FMacro%2FPull%2FPhysical)
- [push of physical elements](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2F%5FMacro%2FPush%2FPhysical)
- [pull of structural/physical panels](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2F%5FMacro%2FPull%2FStructure)

### Changelog

Push/pull methods for physical and analytical elements were refactored. Main assumptions:
- both Physical and Discipline-relevant elements can be pulled
- only Physical elements can be pushed - element types are matched by name, e.g. to push a HEB300 column one need to match the column property name with relevant family type name in Revit
- when pulling a Physical element, a Physical material is created with discipline-relevant fragments
- when pulling a Discipline-relevant element, only a discipline-relevant fragment is pulled
- structural material convert reads "Material Grade" parameter of an element to match it with the library in order to assign the appropriate grade of Structural material fragment
- when pulling a physical element, physical geometry of an element is taken
- when pulling an analytical element, first the analytical geometry is taken:
    * if not null then use it with a warning that analytical geometry is used
    * if null then us the physical geometry
- materials and family types cannot be pushed/updated to Revit - it was like that before and will require a solid batch of work to make sure that is not dangerous

### Additional comments

PR labeled as bug as Revit_Toolkit does not work as expected atm.

### CODE REVIEW

**Comment by pawelbaran** [comment]
@ZiolkowskiJakub please be aware that retrieving the Physical materials from `RefObjects` does not make sense any more: the material is added to `RefObjects` before its properties are assigned.

**Comment by pawelbaran** [comment]
Adding structural guys to the review - would be awesome if you could try to run some arbitrary physical & analytical push/pull tests/models. I am working on a more formalized one, should be ready in the coming days.

Known issues to be resolved before merging:
- [ ] formalized test files for physical & analytical push/pull
- [ ] cleanup of warnings
- [ ] atm justification of bars gets reset on push - the elements are placed in right locations, but their y and z justification/offset parameters in Revit are set to zero
- [ ] code cleanup (removal of irrelevant commented out stuff)
- [ ] anything critical/reachable within a few hours that will be requested

Aiming to merge in the 2nd half of this week.

File: Engine_Revit_UI/Query/Layer.cs [hunk 1]
```diff
@@ -33,66 +33,23 @@ public static partial class Query
...
-
-            aMaterial.Name = aMaterial_Revit.Name;
-
-            IEnvironmentMaterial aMaterialProperties = Convert.ToBHoM(aMaterial_Revit, pullSettings) as IEnvironmentMaterial;
-
-            aMaterial.Properties.Add(aMaterialProperties);
-
-            aLayer.Material = aMaterial;
+            Autodesk.Revit.DB.Material revitMaterial = doc.GetElement(compoundStructureLayer.MaterialId) as Autodesk.Revit.DB.Material;
+            aLayer.Material = revitMaterial.ToBHoMEmptyMaterial(pullSettings);
```
Comment (ZiolkowskiJakub) [inline Engine_Revit_UI/Query/Layer.cs]: Can we set Layer name after Material name?

File: Engine_Revit_UI/Convert/Physical/ToRevit/FamilyInstance.cs [hunk 2]
```diff
@@ -132,26 +131,47 @@ private static FamilyInstance ToRevitFamilyInstance_Column(this IFramingElement
                 Parameter aParameter = aFamilyInstance.get_Parameter(BuiltInParameter.STRUCTURAL_BEND_DIR_ANGLE);
                 if (aParameter != null && !aParameter.IsReadOnly)
                     aParameter.Set(orientationAngle);
+
+                //TODO: if the material does not get assigned an error should be thrown?
+                if (barProperty.Material != null && barProperty.Material.CustomData.ContainsKey("Revit_elementId"))
```
[the remark this answers is not in the document]
Comment (pawelbaran) [reply Engine_Revit_UI/Convert/Physical/ToRevit/FamilyInstance.cs]: Done

File: Engine_Revit_UI/Convert/Physical/ToRevit/FamilyInstance.cs [hunk 3]
```diff
@@ -132,26 +131,47 @@ private static FamilyInstance ToRevitFamilyInstance_Column(this IFramingElement
...
+                {
+                    Autodesk.Revit.DB.Material material = document.GetElement(new ElementId((int)barProperty.Material.CustomData["Revit_elementId"])) as Autodesk.Revit.DB.Material;
```
[the remark this answers is not in the document]
Comment (pawelbaran) [reply Engine_Revit_UI/Convert/Physical/ToRevit/FamilyInstance.cs]: Done

File: Engine_Revit_UI/Convert/Physical/ToRevit/FamilyInstance.cs [hunk 4]
```diff
@@ -274,14 +297,21 @@ private static FamilyInstance ToRevitFamilyInstance_Framing(this IFramingElement
...
```
[the remark this answers is not in the document]
Comment (pawelbaran) [reply Engine_Revit_UI/Convert/Physical/ToRevit/FamilyInstance.cs]: Done

File: Engine_Revit_UI/Convert/Physical/ToRevit/FamilyInstance.cs [hunk 5]
```diff
@@ -274,14 +297,21 @@ private static FamilyInstance ToRevitFamilyInstance_Framing(this IFramingElement
                 Parameter aParameter = aFamilyInstance.get_Parameter(BuiltInParameter.STRUCTURAL_BEND_DIR_ANGLE);
                 if (aParameter != null && !aParameter.IsReadOnly)
                     aParameter.Set(orientationAngle);
+
+                //TODO: if the material does not get assigned an error should be thrown?
+                if (barProperty.Material != null && barProperty.Material.CustomData.ContainsKey("Revit_elementId"))
```
[the remark this answers is not in the document]
Comment (pawelbaran) [reply Engine_Revit_UI/Convert/Physical/ToRevit/FamilyInstance.cs]: Done

File: Engine_Revit_UI/Convert/Structure/ToBHoM/MaterialFragment.cs [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+            oM.Geometry.Vector thermalExpansionCoeff = BH.Engine.Geometry.Create.Vector(UnitUtils.ConvertFromInternalUnits(structuralAsset.ThermalExpansionCoefficient.X, DisplayUnitType.DUT_INV_CELSIUS), UnitUtils.ConvertFromInternalUnits(structuralAsset.ThermalExpansionCoefficient.Y, DisplayUnitType.DUT_INV_CELSIUS), UnitUtils.ConvertFromInternalUnits(structuralAsset.ThermalExpansionCoefficient.Z, DisplayUnitType.DUT_INV_CELSIUS));
+            oM.Geometry.Vector poissonsRatio = BH.Engine.Geometry.Create.Vector(structuralAsset.PoissonRatio.X, structuralAsset.PoissonRatio.Y, structuralAsset.PoissonRatio.Z);
+            oM.Geometry.Vector shearModulus = BH.Engine.Geometry.Create.Vector(UnitUtils.ConvertFromInternalUnits(structuralAsset.ShearModulus.X, DisplayUnitType.DUT_PASCALS), UnitUtils.ConvertFromInternalUnits(structuralAsset.ShearModulus.Y, DisplayUnitType.DUT_PASCALS), UnitUtils.ConvertFromInternalUnits(structuralAsset.ShearModulus.Z, DisplayUnitType.DUT_PASCALS));
+
+            materialFragment.Density = density;
+            materialFragment.DampingRatio = dampingRatio;
+
+            if (materialFragment is BH.oM.Structure.MaterialFragments.Aluminium)
```
[the remark this answers is not in the document]
Comment (pawelbaran) [reply Engine_Revit_UI/Convert/Structure/ToBHoM/MaterialFragment.cs]: Yes, correct. The values are converted at the beginning of this method.

File: Engine_Revit_UI/Query/Layer.cs (hunk 1 above)
Comment (pawelbaran) [reply Engine_Revit_UI/Query/Layer.cs]: Which name would you like to assign to the layer? `Autodesk.Revit.DB.CompoundStructureLayer` does not have one :<

**Comment by pawelbaran** [comment]
Framing adjustment/offset fixed for straight bars - the code to be cleaned up tomorrow.

**Comment by JosefTaylor** [comment]
#171 is not solved by this PR, as that is an issue with the US template- nobody told us we had to set the section type parameter for our column/framing families. I'll edit the issue to reflect this.

**Comment by pawelbaran** [comment]
@JosefTaylor thanks for having a look. Could you please place a model representing #171 (preferably with problem pointed out by @enarhi) on SharePoint? I have no experience with the US template (I do not even know where to find it on the server).

BTW, section type parameter is also an issue in case of some of the UK/Europe families - we need to build a way to generate these automatically based on the verified section library.

**Comment by ZiolkowskiJakub** [comment]
Hi @pawelbaran @michaldengusiak,

I updated PR to include Openings for all ISurfaces (Wall, Floor, Roof etc.). The rules are as follow:

- If you create opening in profile of Revit element then it should go to the Location of the ISurface (InternalBoundaries)

- If the opening is created by hosted object such as Window or Door then it will go to ISurface.Openings

**Comment by al-fisher** [comment]
Hi @pawelbaran @ZiolkowskiJakub - initial testing on pull `physical.elements` is looking generally good on various models I am using - getting largely expected results.

However on some elements I am getting the following warning:

> Could not find Revit Thermal Asset for BHoM Object. BHoM Guid: ea404040-b933-4a1a-8b64-7b919ef6282c

Is that expected?

**Comment by pawelbaran** [comment]
> However on some elements I am getting the following warning:
> 
> Could not find Revit Thermal Asset for BHoM Object. BHoM Guid: ea404040-b933-4a1a-8b64-7b919ef6282c
> 
> Is that expected?

Yeah, this means that the thermal properties (in BEnv understanding, not the thermal expansion coefficient) of the material are not set in Revit.

**Comment by pawelbaran** [comment]
> Sure thing @pawelbaran:
> https://burohappold.sharepoint.com/:f:/s/BHoM/Em2GjEWXzx1Os0k8uPhXDjIBhjiu_Po8aJ4hXoDL4r16LQ?e=NDO002
> 
> The issue might not be exactly as @enarhi described, but they do come in invalid.

The profiles that do not work are:
- tapered
- double profiles

These are still not well supported (in BHoM as such, not only in Revit_TK). The rest goes out fine, apart from the fact that without the _Section Shape_ param we cannot identify the shape, so a freeform profile is returned. BTW, I just noticed that the freeform section profile is rotated 90 deg - will fix it before merging, so please do not worry.

**Comment by JosefTaylor** [comment]
Hmm, it seemed to me like the freeform shapes don't come out at all- the profile has edges, but these edges won't preview or convert to GH curves. Am I doing something wrong?

Tapered and doubles not working is expected

**Comment by al-fisher** [comment]
> Yeah, this means that the thermal properties (in BEnv understanding, not the thermal expansion coefficient) of the material are not set in Revit.

OK - cool. Makes sense. Thanks
I wonder if tweaking the warning message to include reference to `Environment` might help clarity?

``` C#
string aMessage = "Could not find Revit Thermal Asset for BH.oM.Environment Fragment.";
```

Something like the above? What do people think?

We do also need a way of handling/displaying 100s of the same warning easily to the user without scaring them! An issue for another time though

**Comment by pawelbaran** [comment]
I have no issues with converting the edges to GH curves (see the screenshot below, script uploaded to [SharePoint](http://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?id=%2Fsites%2FBHoM%2F02_Current%2F12_Scripts%2F01_Test%20Scripts%2FRevit_Toolkit%2FRevit_Toolkit-Issue171-PullUSTemplateFramingFamilies&p=true&originalPath=aHR0cHM6Ly9idXJvaGFwcG9sZC5zaGFyZXBvaW50LmNvbS86Zjovcy9CSG9NL0VtMkdqRVdYengxT3Mwazh1UGhYRGpJQmhqaXVfUG84YUo0aFhvREw0cjE2TFE_cnRpbWU9RkwyOXpnNUIxMGc)). The GH preview works only on simple cuboids I think, so would not be concerned about that. The question is whether you can push the freeform sections to any FE software without converting them to a section shape...

image

**Comment by pawelbaran** [comment]
Agreed, @al-fisher - the warnings need to be cleaned up. I am leaving this to the end, when the functionality is OK. Hopefully will get it more or less covered before merging :-)

**Comment by IsakNaslundBh** [comment]
Small thing found so far testing on the structural sample model shipped with Revit:

- Pulling as Framing some of the unsupported section types are returned with a null profile.
- Pulling as Bars they are returned with a SteelSection with an empty profile.

Snip of the affected family:

image

This is not super critical, as the family is unsupported so far, but would be good to align.

**Comment by JosefTaylor** [comment]
At the moment, you cannot push freeform sections to ETABS or SAP2000 because the 'section designer' package in those softwares is difficult. The best workaround is to use the name of the section to pick a catalogue section, or to create a dummy section with the name, and push that to the FE package- if a section with the same name is defined in the model, it will be used.

**Comment by pawelbaran** [comment]
I resolved as many issues as possible, columns are finished - test files added to the description of this PR. Tomorrow I will go on with framing + need to add openings to BH.oM.Structure.Panel pull once @ZiolkowskiJakub is done on his end.

**Comment by al-fisher** [comment]
Thanks @pawelbaran great effort - I will pick up those test files and give them a spin.

Can all please provide feed back on this thread tomorrow ready for _end of morning UTC_ deadline

Thanks

**Comment by pawelbaran** [comment]
Pull of structural panels added, test files added to the SharePoint (please see PR description).

**Comment by ZiolkowskiJakub** [comment]
Hi @al-fisher, @JosefTaylor, @IsakNaslundBh , @michaldengusiak , @FraserGreenroyd,

Can somebody please run some quick test and approve PR? Test files included in PR description.

**Comment by awakeman** [comment]
I'm having a quick look as I have the capacity.

**Comment by al-fisher** [comment]
Tests here also looking very good. 👍 

My only comment is now minor and on warnings again! 🐒 
Raising issue to pick up later - enable discussion https://github.com/BHoM/Revit_Toolkit/issues/380 - however my simple proposal is to prefix all warnings when pulling Revit Physical with the discipline the warning is relating to.

**Comment by ZiolkowskiJakub** [comment]
@IsakNaslundBh,

Any issues from your further investigation? I need to close this PR ASAP.

Thanks in advance

**Comment by ZiolkowskiJakub** [comment]
Hi.. I am closing this PR... Please rise separate Issue If you find any bug related to this PR.

**Comment by FraserGreenroyd** [comment]
Test runs were looking good from my side just for completeness :smile: great work @pawelbaran and @ZiolkowskiJakub !

**Pull request merged**